File: cpp/code/climbing-stairs-2.cpp
// Climbing Stairs
// 数学公式，时间复杂度O(1)，空间复杂度O(1)
class Solution {
    public:
    int climbStairs(int n) {
        const double s = sqrt(5);
        return floor((pow((1+s)/2, n+1) + 
            pow((1-s)/2, n+1))/s + 0.5);
    }
};

File: java/code/climbing-stairs-2.java
// Climbing Stairs
// 数学公式，时间复杂度O(1)，空间复杂度O(1)
public class Solution {
    public int climbStairs(int n) {
        final double s = Math.sqrt(5);
        return (int)Math.floor((Math.pow((1+s)/2, n+1) + 
            Math.pow((1-s)/2, n+1))/s + 0.5);
    }
};